src/indexes/text/flat_position_map.cc

```diff
@@ -0,0 +1,590 @@
+
+  // For large position lists, use binary search encoding
+  if (num_positions > 128) {
+    return EncodingScheme::BINARY_SEARCH;
+  }
+
+  // Check if simple encoding is sufficient
+  // Simple encoding: single field, all positions fit in 1 byte
+  if (num_text_fields == 1) {
+    Position max_pos = position_map.rbegin()->first;
+    if (max_pos < 256) {
+      return EncodingScheme::SIMPLE;
+    }
+  }
+
+  // Default to expandable encoding
+  return EncodingScheme::EXPANDABLE;
+}
+
+// Serialize using SIMPLE encoding
+FlatPositionMap SerializeSimple(
+    const std::map<Position, std::unique_ptr<FieldMask>>& position_map) {
+  uint32_t num_positions = position_map.size();
+
+  // Header (4 bytes) + position data (1 byte per position)
+  size_t total_size = 4 + num_positions;
+  char* flat_map = static_cast<char*>(malloc(total_size));
+  CHECK(flat_map != nullptr) << "Failed to allocate FlatPositionMap";
+
+  // Write header
+  uint32_t header = (0 << 0) |  // Standard header
+                    (static_cast<uint32_t>(EncodingScheme::SIMPLE) << 1) |
+                    (num_positions << 3);
+  WriteUint32(flat_map, header);
+
+  // Write positions (delta encoded)
+  char* data_ptr = flat_map + 4;
+  Position prev_pos = 0;
+  for (const auto& [pos, field_mask] : position_map) {
+    uint32_t delta = pos - prev_pos;
+    CHECK(delta < 256) << "Delta too large for simple encoding";
+    *data_ptr++ = static_cast<char>(delta);
+    prev_pos = pos;
+  }
+
+  return flat_map;
+}
+
+// Serialize using EXPANDABLE encoding
+FlatPositionMap SerializeExpandable(
+    const std::map<Position, std::unique_ptr<FieldMask>>& position_map,
+    size_t num_text_fields) {
+  uint32_t num_positions = position_map.size();
+  uint8_t field_bytes = GetFieldMaskBytes(num_text_fields);
+
+  // Estimate size (conservative upper bound)
+  // Each entry: up to 4 bytes for position + at least 1 byte for field
+  // mask/terminator
+  size_t estimated_size =
+      4 + (num_positions * (4 + (field_bytes > 0 ? field_bytes : 1)));
+  char* flat_map = static_cast<char*>(malloc(estimated_size));
+  CHECK(flat_map != nullptr) << "Failed to allocate FlatPositionMap";
+
+  // Write header (will update later with actual size)
+  uint32_t header = (0 << 0) |  // Standard header
+                    (static_cast<uint32_t>(EncodingScheme::EXPANDABLE) << 1) |
+                    (num_positions << 3);
+  WriteUint32(flat_map, header);
+
+  // Write position and field data
+  char* data_ptr = flat_map + 4;
+  Position prev_pos = 0;
+
+  for (const auto& [pos, field_mask] : position_map) {
+    uint32_t delta = pos - prev_pos;
+
+    // Encode position delta with variable bytes
+    // First bit: 0 = position byte, 1 = field byte
+    // Remaining 7 bits: value
+    while (delta >= 128) {
+      *data_ptr++ = static_cast<char>(0x00 | (delta & 0x7F));
+      delta >>= 7;
+    }
+    *data_ptr++ = static_cast<char>(0x00 | delta);
+
+    // Encode field mask
```

Comment:
Weren't we going to delta-encode the field mask?

Reply:
I thought it was in the plan to only add a field mask when it changed. If a bunch of positions in a row are for the same field(s), then no need to repeat the same field mask over and over. One consideration for this would be to add the field mask in the partition info for the `BinarySearch` encoding